### ISSUE TITLE
Inconsistent behavior of IN predicate with constant vs. non-constant in-list

### ISSUE THREAD

**Issue by kagamiori** [prestodb/presto#23332]
The IN predicate returns different results for two queries with the same operands, except one query uses constant in-list and the other uses non-constant in-list. I'd expect these two queries to both return false.

```
select c0 in (c1, c2) from (values (array[array[null]], array[array[1,2]], array[array[2,1]])) t(c0, c1, c2); -- false

select c0 in (array[array[1,2]], array[array[2,1]]) from (values (array[array[null]])) t(c0); -- NULL
```

In addition, the following two queries return different results. I'd expect them to return the same result given the semantic of the queries. (These two queries return different results because the first query below is rewritten to not use IN predicate but do a equal-comparison `c0 = array[array[1,2]]`.)

```
select c0 in (array[array[1,2]]) from (values (array[array[null]])) t(c0); -- false

select c0 in (array[array[1,2]], array[array[1,2]]) from (values (array[array[null]])) t(c0); -- NULL
```

## Your Environment

* Presto version used: 0.289-edge8.1
* Storage (HDFS/S3/GCS..):
* Data source and connector used:
* Deployment (Cloud or On-prem):
* [Pastebin](https://pastebin.com/) link to the complete debug logs:

**Comment by mbasmanova** [comment]
CC: @kaikalur

**Comment by kaikalur** [comment]
Yeah - one is const folded and the other one not I think. See: #23245 maybe that will make it consistent

**Comment by rschlussel** [comment]
I'd expect these queries to all return null (match semantics of when a null is used directly)
```
presto:di> select null in (2, 1);
 _col0                
-------
 NULL  
(1 row)
```

**Comment by kagamiori** [comment]
> I'd expect these queries to all return null (match semantics of when a null is used directly)
> 
> ```
> presto:di> select null in (2, 1);
>  _col0                
> -------
>  NULL  
> (1 row)
> ```
> 
> so i think there is a bug in the constant folding.

Hi @rschlussel, I think (at least currently) the NULL handling in comparisons for complex types is slightly different from that in comparisons for primitive types. E.g., two arrays of different sizes are distinguishable even if they contain NULLs. So I expected all queries above to return false.
``` 
select array[null] = array[null, null] -- false
```

**Comment by rschlussel** [comment]
ah, I see, thanks for explaining.  Because they have a different number of elements, they should return false.
